src/main/java/io/vertx/pgclient/codec/decoder/PgMessageDecoder.java

```diff
@@ -0,0 +1,424 @@
+          break;
+
+        case ROUTINE:
+          responseMessage.setRoutine(readCString(data, UTF_8));
+          break;
+
+        case SCHEMA:
+          responseMessage.setSchema(readCString(data, UTF_8));
+          break;
+
+        case TABLE:
+          responseMessage.setTable(readCString(data, UTF_8));
+          break;
+
+        case COLUMN:
+          responseMessage.setColumn(readCString(data, UTF_8));
+          break;
+
+        case DATA_TYPE:
+          responseMessage.setDataType(readCString(data, UTF_8));
+          break;
+
+        case CONSTRAINT:
+          responseMessage.setConstraint(readCString(data, UTF_8));
+          break;
+
+        default:
+          readCString(data, UTF_8);
+          break;
+      }
+    }
+    out.add(responseMessage);
+  }
+
+  private void decodeAuthentication(ByteBuf data, List<Object> out) {
+
+    final int AUTHENTICATION_OK = 0;
+    final int AUTHENTICATION_KERBEROS_V5 = 2;
+    final int AUTHENTICATION_CLEARTEXT_PASSWORD = 3;
+    final int AUTHENTICATION_MD5_PASSWORD = 5;
+    final int AUTHENTICATION_SCM_CREDENTIAL = 6;
+    final int AUTHENTICATION_GSS = 7;
+    final int AUTHENTICATION_GSS_CONTINUE = 8;
+    final int AUTHENTICATION_SSPI = 9;
+
+    int AUTHENTICATION_TYPE = data.readInt();
+    switch (AUTHENTICATION_TYPE) {
+      case AUTHENTICATION_OK: {
+        out.add(new AuthenticationOkMessage());
+      }
+      break;
+      case AUTHENTICATION_MD5_PASSWORD: {
+        byte[] salt = new byte[4];
+        data.readBytes(salt);
+        out.add(new AuthenticationMD5PasswordMessage(salt));
+      }
+      break;
+      case AUTHENTICATION_CLEARTEXT_PASSWORD: {
+        out.add(new AuthenticationClearTextPasswordMessage());
+      }
+      break;
+      case AUTHENTICATION_KERBEROS_V5:
+      case AUTHENTICATION_SCM_CREDENTIAL:
+      case AUTHENTICATION_GSS:
+      case AUTHENTICATION_GSS_CONTINUE:
+      case AUTHENTICATION_SSPI:
+      default:
+        throw new UnsupportedOperationException("Authentication type is not supported in the client");
+    }
+  }
+
+  private void decodeCommandComplete(ByteBuf data, List<Object> out) {
+
+    final String INSERT = "INSERT";
+    final String SELECT = "SELECT";
+    final String UPDATE = "UPDATE";
+    final String DELETE = "DELETE";
+    final String MOVE = "MOVE";
+    final String FETCH = "FETCH";
+    final String COPY = "COPY";
+    final String CREATE = "CREATE";
+    final String DROP = "DROP";
+    final String ALTER = "ALTER";
+    final String DECLARE = "DECLARE";
+    final String CLOSE = "CLOSE";
+    final String PREPARE = "PREPARE";
+    final String COMMIT = "COMMIT";
+    final String ROLLBACK = "ROLLBACK";
+    final String DEALLOCATE = "DEALLOCATE";
+    final String TRUNCATE = "TRUNCATE";
+    final String LOCK = "LOCK";
+    final String GRANT = "GRANT";
+    final String REVOKE = "REVOKE";
+
+
+    String tag = readCString(data, UTF_8);
+    String[] parts = tag.split(" ");
```

Comment:
avoir array creation as it's used below with `command` that can be a `StringBuilder`, we can also avoid the `tag` creation` by just using an index.

Overall we can even use a `ByteBuf` instead a `StringBuilder`